src/Physics/babylon.physicsImpostor.ts

```diff
@@ -330,13 +330,22 @@ module BABYLON {
             }
         }
 
+        /**
+         * Legacy collision detection event support
+         */
+        public onCollideEvent: (collider:BABYLON.PhysicsImpostor, collidedWith:BABYLON.PhysicsImpostor) => void = null;
+
         //event and body object due to cannon's event-based architecture.
         public onCollide = (e: { body: any }) => {
-            if (!this._onPhysicsCollideCallbacks.length) return;

```

Comment:
No worry for the particle code. We prefer having just one feature per PR just in case we need to revert back.
Regarding your code, can you please change to this (We clearly need to be utterly cautious when dealing with performance intensive code):
`
                public onCollide = (e: { body: any }) => {

                if (!this._onPhysicsCollideCallbacks.length && !this.onCollideEvent) {
                   return;
                }

            var otherImpostor = this._physicsEngine.getImpostorWithPhysicsBody(e.body);
            if (otherImpostor) {
                // Legacy collision detection event support
                if (this.onCollideEvent) {
                    this.onCollideEvent(this, otherImpostor);
                }
                this._onPhysicsCollideCallbacks.filter((obj) => {
                    return obj.otherImpostors.indexOf(otherImpostor) !== -1
                }).forEach((obj) => {
                    obj.callback(this, otherImpostor);
                })
            }
        }`